launcher/src/main/java/org/apache/spark/launcher/AbstractLauncher.java

```diff
@@ -0,0 +1,600 @@
+  public T setJavaHome(String path) {
+    checkNotNull(path, "path");
+    this.javaHome = path;
+    return THIS;
+  }
+
+  /** Set a custom Spark installation location for the application. */
+  public T setSparkHome(String path) {
+    checkNotNull(path, "path");
+    this.sparkHome = path;
+    return THIS;
+  }
+
+  /** Set a custom properties file with Spark configuration for the application. */
+  public T setPropertiesFile(String path) {
+    checkNotNull(path, "path");
+    this.propertiesFile = path;
+    return THIS;
+  }
+
+  /** Set a single configuration value for the application. */
+  public T setConf(String key, String value) {
+    checkNotNull(key, "key");
+    checkNotNull(value, "value");
+    checkArgument(key.startsWith("spark."), "'key' must start with 'spark.'");
+    conf.put(key, value);
+    return THIS;
+  }
+
+  /**
+   * Launchers should implement this to create the command to be executed.
+   */
+  protected abstract List<String> buildLauncherCommand() throws IOException;
+
+  protected Properties loadPropertiesFile() throws IOException {
+    Properties props = new Properties();
+    File propsFile;
+    if (propertiesFile != null) {
+      propsFile = new File(propertiesFile);
+      checkArgument(propsFile.isFile(), "Invalid properties file '%s'.", propertiesFile);
+    } else {
+      String confDir = getenv("SPARK_CONF_DIR");
+      if (confDir == null) {
+        confDir = join(File.separator, getSparkHome(), "conf");
+      }
+      propsFile = new File(confDir, "spark-defaults.conf");
+    }
+
+    if (propsFile.isFile()) {
+      FileInputStream fd = null;
+      try {
+        fd = new FileInputStream(propsFile);
+        props.load(new InputStreamReader(fd, "UTF-8"));
+      } finally {
+        if (fd != null) {
+          try {
+            fd.close();
+          } catch (IOException e) {
+            // Ignore.
+          }
+        }
+      }
+    }
+
+    return props;
+  }
+
+  protected String getSparkHome() {
+    String path = first(sparkHome, getenv("SPARK_HOME"));
+    checkState(path != null,
+        "Spark home not found; set it explicitly or use the SPARK_HOME environment variable.");
+    return path;
+  }
+
+  protected List<String> createJavaCommand() throws IOException {
+    List<String> cmd = new ArrayList<String>();
+    if (javaHome == null) {
+      cmd.add(join(File.separator, System.getProperty("java.home"), "..", "bin", "java"));
+    } else {
+      cmd.add(join(File.separator, javaHome, "bin", "java"));
+    }
+
+    // Don't set MaxPermSize for Java 8 and later.
+    String[] version = System.getProperty("java.version").split("\\.");
+    if (Integer.parseInt(version[0]) == 1 && Integer.parseInt(version[1]) < 8) {
+      cmd.add("-XX:MaxPermSize=128m");
+    }
+
+    // Load extra JAVA_OPTS from conf/java-opts, if it exists.
+    File javaOpts = new File(join(File.separator, getSparkHome(), "conf", "java-opts"));
```

Comment:
I see. I think that's just an omission in the original code. Let's just fix it here